### PR TITLE
feat(parser): handle optional chaining in scope analysis

### LINKED ISSUE #1090: [no-restricted-globals] False positive with optional chaining
I wasn't quite sure which should sub-project was appropriate for this issue. I am aware that strictly ts 3.7.0-beta isn't supported while TS 3.7 syntax is, but out of caution I thought to raise the issue as I couldn't see any other reports.

**What code were you trying to parse?**

```ts
// foobar.ts
const a = foo?.bar?.name ?? "foobar"
```

**What did you expect to happen?**

That eslint would not emit any warnings nor errors

**What actually happened?**

Eslint emitted the following error: 
`Unexpected use of 'name'  no-restricted-globals`

This does not happen for all properties, but so far I've only been able to reproduce with properties called `name`.

**Versions**

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/parser` | `2.4.0` |
| `TypeScript`                           | `3.7.0-beta` (15 Oct 2019) |
| `ESLint`                    | `6.5.1` |
| `node`                                 | `12.11.1` |
| `npm`                                  | `6.11.3` |
| `yarn`                                  | `1.19.1` |

### LINKED ISSUE #1104: [no-use-before-define] False positive with optional chaining
**Repro**

```json
{
  "rules": {
    "@typescript-eslint/no-use-before-define": ["error"]
  }
}
```

```ts
const updatedAt = data?.updatedAt;
//                      ^^^^^^^^^
// 'updatedAt' was used before it was defined. (@typescript-eslint/no-use-before-define)
```

**Expected Result**
Would not produce any errors.

**Actual Result**
Produces an error:

```ts
// 'updatedAt' was used before it was defined. (@typescript-eslint/no-use-before-define)
```

**Additional Info**
This issue seems to only happen when the variable name is the same as the property name. For instance, the code below does not show the warning:

```ts
const timestamp = data?.updatedAt;
```

**Versions**

package | version
-- | --
@typescript-eslint/parser | `2.4.0`
TypeScript | `3.7.0-beta` (15 Oct 2019)
ESLint | `6.5.1`
node | `12.11.1`
npm | `6.11.3`
yarn | `1.19.1`

### LINKED ISSUE #1116: [no-undef] False positive with optional chaining
Hi there, 

Just trying the latest eslint-plugin (2.4.0) with optional chaining and this code:

```JavaScript
const a: { x: number; y: number; z?: number} = { x: 1, y: 2 };
const b = a?.z ?? 3;

console.log(b);
```
Gives a no-undef error on `.z`

```Javascript
module.exports = {
	'globals': {
		'Class': true,
		'SyntheticMouseEvent': true,
		'SyntheticChangeEvent': true,
		'TimeoutID': true,
		'IntervalID': true,
		'Blob': true
	},

	'extends': [
		'eslint-config-airbnb',
		'plugin:@typescript-eslint/recommended'
	],

	parser: '@typescript-eslint/parser',
	plugins: [
		'@typescript-eslint',
		'@typescript-eslint/eslint-plugin',
		'react'
	],

	'env': {
		'browser': true,
		'node': true,
		'es2017': true,
		'mocha': true
	},
	'rules': {
		'no-bitwise': 'off',
		'no-console': 'off',
		'spaced-comment': 'off',
		'no-param-reassign': 'off',
		'no-return-assign': 'warn',
		'no-nested-ternary': 'off',
		'no-restricted-globals': 'off',
		'class-methods-use-this': 'off',
		'sort-keys': 0,
		'indent': [
			'error',
			'tab',
			{
				'SwitchCase': 1
			}
		],
		'no-confusing-arrow': 'off',
		'implicit-arrow-linebreak': 'off',
		'lines-between-class-members': 'off',
		'prefer-destructuring': 'off',

		'arrow-body-style': 'off',
		'operator-linebreak': 'off',
		'no-underscore-dangle': 0,
		'no-tabs': 0,
		'no-shadow': 0,
		'no-useless-escape': 0,
		'padded-blocks': 0,
		'space-in-parens': 0,
		'template-curly-spacing': [
			'error',
			'always'
		],
		'object-curly-newline': 'off',
		'comma-dangle': [
			'error',
			'never'
		],
		'no-trailing-spaces': 'off',
		'no-unused-vars': 'warn',
		'no-unused-expressions': 'off',
		'linebreak-style': [
			'error',
			'unix'
		],
		'max-len': [
			'error',
			400,
			2,
			{
				'ignoreUrls': true,
				'ignoreComments': false,
				'ignoreRegExpLiterals': true,
				'ignoreStrings': true,
				'ignoreTemplateLiterals': true
			}
		],

		'import/prefer-default-export': 0,
		'import/no-unresolved': 0,
		'import/no-cycle': 'off',
		'import/no-extraneous-dependencies': 'off',
		'import/no-webpack-loader-syntax': 0,

		'react/prefer-stateless-function': 'off',
		'react/no-multi-comp': 'off',
		'react/jsx-boolean-value': 'off',
		'react/jsx-closing-tag-location': 'off',
		'react/jsx-closing-bracket-location': 'off',
		'react/jsx-curly-spacing': [
			'error',
			{
				'when': 'always'
			}
		],
		'react/jsx-filename-extension': [
			'warn',
			{
				'extensions': [
					'.ts',
					'.tsx'
				]
			}
		],
		'react/jsx-indent-props': [
			'error',
			'tab'
		],
		'react/jsx-one-expression-per-line': 'off',
		'react/jsx-indent': [
			'error',
			'tab'
		],
		'react/button-has-type': 'off',
		'react/no-unescaped-entities': 'off',
		'react/prop-types': 0,
		'react/require-default-props': 0,
		'react/destructuring-assignment': 'off',
		'react/sort-comp': 'off',
		'react/state-in-constructor': 'off',
		'react/static-property-placement': 'off',
		'react/no-array-index-key': 'off',

		'@typescript-eslint/class-name-casing': 'off',
		'@typescript-eslint/explicit-function-return-type': 'off',
		'@typescript-eslint/ban-ts-ignore': 'off',
		'@typescript-eslint/no-explicit-any': 'off',

		'jsx-a11y/anchor-is-valid': 'off',
		'jsx-a11y/click-events-have-key-events': 'off',
		'jsx-a11y/href-no-hash': 'off',
		'jsx-a11y/label-has-for': 'off',
		'jsx-a11y/mouse-events-have-key-events': 'off',
		'jsx-a11y/no-autofocus': 'off',
		'jsx-a11y/no-noninteractive-element-interactions': 'off',
		'jsx-a11y/no-noninteractive-tabindex': 'off',
		'jsx-a11y/no-static-element-interactions': 'off',
		'jsx-a11y/show-errors': 'off',
		'jsx-a11y/tabindex-no-positive': 'off'
	}
};
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.4.0` |
| `@typescript-eslint/parser`        | `2..4.0` |
| `eslint-config-airbnb` | `18.0.1` |
| `TypeScript`                       | `3.7.0-beta` |
| `ESLint`                           | `6.5.1` |
| `node`                             | `10.16.2` |
| `npm`                              | `6.10.3` |

### PR DESCRIPTION
**Pull request by Nizarius**
Fixes #1090
Fixes #1116
Fixes #1104

### CODE REVIEW

**Comment by Nizarius** [comment]
Guys, can you please give me any update on this PR? It looks like crucial bug but in 4 days no comments (or merging) were provided.

**Comment by bradzacher** [comment]
> The core maintainers work on this in their personal time, so please understand that it may not be possible for them to review your work immediately.

Myself and the other maintainers volunteer our time to this project. We do not always have time every week to review PRs. 

We'll get to it as soon as we can.

File: packages/eslint-plugin/tests/eslint-rules/no-restricted-globals.test.ts [hunk 1]
```diff
@@ -30,6 +30,11 @@ type Handler = (event: string) => any
       `,
       options: ['event'],
     },
+    {
+      code: `
+        const a = foo?.bar?.name ?? "foobar"
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/eslint-rules/no-restricted-globals.test.ts]: Prefer a separate test for nullish coalescing and optional chaining
```suggestion
        const a = foo?.bar?.name
```

File: packages/parser/src/analyze-scope.ts [hunk 2]
```diff
@@ -344,6 +344,17 @@ class Referencer extends TSESLintScope.Referencer<ScopeManager> {
     node.arguments.forEach(this.visit, this);
   }
 
+  /**
+   * Visit optional member expression.
+   * @param node The OptionalMemberExpression node to visit.
+   */
+  OptionalMemberExpression(node: TSESTree.OptionalMemberExpression): void {
```
Comment (bradzacher) [inline packages/parser/src/analyze-scope.ts]: should also add support for `OptionalCallExpression` in this PR
(i.e. pull in your change from #1175)

And add tests accordingly

Comment (Nizarius) [reply packages/parser/src/analyze-scope.ts]: Done

File: packages/eslint-plugin/tests/eslint-rules/no-restricted-globals.test.ts (hunk 1 above)
Comment (Nizarius) [reply packages/eslint-plugin/tests/eslint-rules/no-restricted-globals.test.ts]: Done

**Comment by Nizarius** [comment]
@bradzacher I've added optionalCallExpression to Referencer and added a lot of new tests to these rules

**Comment by deser** [comment]
Let's merge?:)

**Comment by xwchris** [comment]
when the pr can be merged?

**Comment by bradzacher** [comment]
@deser and @xwchris - please don't comment with comments like that.
It serves no purpose, and just creates spam notifications for everyone (i.e. for myself and the other maintainers, or anyone watching this repo).

***Let me be clear: it doesn't bump the priority, it just spams people.***

As I mentioned before, this project is maintained by volunteers. We will get to a review when we're able.

**Comment by davidje13** [comment]
Using the just-published 2.7.0 (which was released after this patch) I'm still seeing "no-restricted-globals" and "@typescript-eslint/no-use-before-define" errors when using `?.` syntax. (#1090 #1104). Is it possible this patch somehow got missed from the released build? I haven't done much investigating yet.

**Pull request merged**